parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java

```diff
@@ -1724,160 +1975,125 @@ public ColumnChunkPageReader readAllPages(
-            if (options.usePageChecksumVerification() && pageHeader.isSetCrc()) {
-              pageBytes = BytesInput.concat(repetitionLevels, definitionLevels, values);
-              verifyCrc(
-                  pageHeader.getCrc(),
-                  pageBytes,
-                  "could not verify page integrity, CRC checksum verification failed");
+
+        @Override
+        public boolean hasNext() {
+          if (!bufferedToFirstDataPage) {
+            bufferNextDataPage();
+          }
+          return nextPage != null;
+        }
+
+        private boolean hasMorePages(long valuesCountReadSoFar, int dataPageCountReadSoFar) {
+          return offsetIndex == null
+              ? valuesCountReadSoFar < descriptor.metadata.getValueCount()
+              : dataPageCountReadSoFar < offsetIndex.getPageCount();
+        }
+
+        private void bufferNextDataPage() {
+          bufferedToFirstDataPage = true;
+          while (true) {
+            if (!hasMorePages(valuesCountReadSoFar, dataPageCountReadSoFar)) {
+              return;
             }
-            DataPageV2 dataPageV2 = new DataPageV2(
-                dataHeaderV2.getNum_rows(),
-                dataHeaderV2.getNum_nulls(),
-                dataHeaderV2.getNum_values(),
-                repetitionLevels,
-                definitionLevels,
-                converter.getEncoding(dataHeaderV2.getEncoding()),
-                values,
-                uncompressedPageSize,
-                converter.fromParquetStatistics(
-                    getFileMetaData().getCreatedBy(), dataHeaderV2.getStatistics(), type),
-                dataHeaderV2.isIs_compressed());
-            // Copy crc to new page, used for testing
-            if (pageHeader.isSetCrc()) {
-              dataPageV2.setCrc(pageHeader.getCrc());
+            try {
+              byte[] pageHeaderAAD = dataPageHeaderAAD;
+              if (null != headerBlockDecryptor) {
+                // Important: this verifies file integrity (makes sure dictionary page had not been
+                // removed)
+                if (null == dictionaryPage && descriptor.metadata.hasDictionaryPage()) {
+                  pageHeaderAAD = AesCipher.createModuleAAD(
+                      aadPrefix,
+                      ModuleType.DictionaryPageHeader,
+                      rowGroupOrdinal,
+                      columnOrdinal,
+                      -1);
+                } else {
+                  int pageOrdinal = getPageOrdinal(dataPageCountReadSoFar);
+                  AesCipher.quickUpdatePageAAD(dataPageHeaderAAD, pageOrdinal);
+                }
+              }
+
+              PageWithHeader pageWithHeader =
+                  chunkStream.readNextPage(chunk, type, headerBlockDecryptor, pageHeaderAAD);
+              if (pageWithHeader == null) {
+                return;
+              }
+
+              if (pageWithHeader.header.type == PageType.DICTIONARY_PAGE) {
```

Comment:
It seems that the header member variable in the new `PageWithHeader` class is only used here for checking page type. Could we get it from the page class? Perhaps add a Page.type() method?